packages/common/http/src/fetch.ts

```diff
@@ -0,0 +1,229 @@
+import {HttpRequest} from './request';
+import {HttpDownloadProgressEvent, HttpErrorResponse, HttpEvent, HttpEventType, HttpHeaderResponse, HttpResponse, HttpSentEvent, HttpStatusCode,} from './response';
+
+const XSSI_PREFIX = /^\)\]\}',?\n/;
+
+/**
+ * Determine an appropriate URL for the response, by checking either
+ * response url or the X-Request-URL header.
+ */
+function getResponseUrl(response: Response): string|null {
+  if (response.url) {
+    return response.url;
+  }
+  // stored as lowercase in the map
+  const xRequestUrl = 'X-Request-URL'.toLocaleLowerCase();
+  return response.headers.get(xRequestUrl);
+}
+
+/**
+ * Uses `fetch` to send requests to a backend server.
+ *
+ * This `FetchBackend` requires the support of the Fetch API which is available on all evergreen
+ * browsers and on NodeJS from v18 onward.
+ *
+ * @see {@link HttpHandler}
+ *
+ * @developerPreview
+ */
+@Injectable()
+export class FetchBackend implements HttpBackend {
+  // We need to bind the native fetch to its context or it will throw an "illegal invocation"
+  private readonly fetchImpl =
+      inject(FetchFactory, {optional: true})?.fetch ?? fetch.bind(globalThis);
+
+  handle(req: HttpRequest<any>): Observable<HttpEvent<any>> {
+    const abortController = new AbortController();
+    // TODO: use `AsyncGenerators` instead of `BehaviorSubject` when we no longer support RXJS 6.x
+    // We can't use it now because from() doesn't accept generators on RxJS 6.
+    const eventStream = new BehaviorSubject<HttpEvent<any>>({type: HttpEventType.Sent});
+
+    return merge(from(this.doRequest(req, abortController.signal, eventStream)), eventStream)
+        .pipe(finalize(() => {
+          // Aborting the fetch call when the observable is unsubscribed
+          abortController.abort();
+        }));
+  }
+
+  private async doRequest(
+      request: HttpRequest<any>, signal: AbortSignal,
+      eventStream: BehaviorSubject<HttpEvent<any>>): Promise<HttpResponse<any>> {
+    const init = this.createRequestInit(request);
+    let response: Response|undefined;
+
+    try {
+      response = await this.fetchImpl(request.url, {signal, ...init});
+    } catch (error: any) {
+      throw new HttpErrorResponse({
+        error,
+        status: error.status ?? 0,
+        statusText: error.statusText,
+        url: request.url,
+        headers: error.headers,
+      });
+    }
+
+    const headers = new HttpHeaders(response.headers);
+    const statusText = response.statusText;
+    const url = getResponseUrl(response) ?? request.url;
+
+    let status = response.status;
+    let body: any = null;
+    if (response.body) {
+      // Read Progess
+      const contentLength = response.headers.get('content-length');
+      const chunks: Uint8Array[] = [];
+      const reader = response.body.getReader();
+      let receivedLength = 0;
+
+      if (request.reportProgress) {
+        eventStream.next(new HttpHeaderResponse({headers, status, statusText, url}));
+      }
```

Comment:
This should probably be moved out of the `response.body` block.

```suggestion

      if (request.reportProgress) {
        eventStream.next(new HttpHeaderResponse({headers, status, statusText, url}));
      }
```